[GCS]Use a separate thread in node failure detector to handle heartbeat

WangTaoTheTonic opened the pull request:
<!-- Thank you for your contribution! Please review https://github.com/ray-project/ray/blob/master/CONTRIBUTING.rst before opening a pull request. -->

## Why are these changes needed?

Now in gcs node manager the node failure detector and heartbeat handler were scheduled periodically, by a timer which use the same io service with gcs main thread. If tasks in gcs main thread is too busy, the detector/heartbeat handler might not get a chance to be scheduled, which is the reason some nodes would be treated as dead(which is actually not).

We've found some cases in our workload. 

In this PR, we start a single io service for node failure detector to make it separated with the main thread.

## Related issue number

<!-- For example: "Closes #1234" -->

## Checks

- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [x] I've included any doc changes needed for https://docs.ray.io/en/latest/.
- [x] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failure rates at https://ray-travis-tracker.herokuapp.com/.
- Testing Strategy
   - [x] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested (please justify below)

AmplabJenkins commented:
Can one of the admins verify this patch?

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Ray-PRB/28200/
Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Ray-PRB/28251/
Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Ray-PRB/28249/
Test FAILed.

rkooo567 commented:
This is great! I think this is the bug that was reported many times before, but we could not reproduce.

rkooo567 commented on src/ray/gcs/gcs_server/gcs_server_main.cc:
```diff
@@ -85,5 +92,5 @@ int main(int argc, char *argv[]) {
   gcs_server.Start();
 
-  main_service.run();
+  signal_service.run();
```
Don't we need `boost::asio::io_service::work work` for this?

WangTaoTheTonic replied:
It is definitely needed. I just renamed the variable so could not found it was missed in original codes.

Nice catch!

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Ray-PRB/28322/
Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Ray-PRB/28397/
Test PASSed.

rkooo567 commented on src/ray/gcs/gcs_server/gcs_server.cc:
```diff
@@ -29,12 +29,15 @@ namespace ray {
                   config.grpc_server_thread_num),
-      client_call_manager_(main_service) {}
+      client_call_manager_(*io_services[0]) {
+  RAY_CHECK(io_services.size() >= 2);
```
I think there's no way that this can be bigger than 2 now right? Let's just have a check == 2

WangTaoTheTonic replied:
ok

rkooo567 commented on src/ray/gcs/gcs_client/test/service_based_gcs_client_test.cc:
```diff
@@ -498,8 +488,8 @@ class ServiceBasedGcsClientTest : public ::testing::Test {
   std::unique_ptr<gcs::GcsServer> gcs_server_;
-  std::unique_ptr<std::thread> server_io_service_thread_;
-  std::unique_ptr<boost::asio::io_service> server_io_service_;
+  size_t io_service_num_{2};
```
Can we use a constant instead of 2? Maybe DEFAULT_GCS_NUM_IO_SERVICES with some comment? (Like we need a separate io service to avoid heartbeat delay)

WangTaoTheTonic replied:
added in constants.h.

rkooo567 commented on src/ray/gcs/gcs_server/gcs_server_main.cc:
```diff
@@ -54,7 +55,16 @@ int main(int argc, char *argv[]) {
+  io_service_pool->Run();
+
+  // IO Service for signals handler.
+  boost::asio::io_service signal_service;
```
Can we just use a main thread as a main_service (and don't have a separate signal service) & just an additional io service only for node_manager? It seems a little weird we have io service that is only for signal.

WangTaoTheTonic replied:
I've tried reuse main_service or one of main_services for signal handler but there will be a deadlock since we join the threads in signal handler, which also uses the thread which will be joined.

rkooo567 replied:
I probably didn't understand here. If you use main_service on the main thread, there's no need to join them right? What I am saying here is 

```
boost::asio::io_service main_service;
...
boost::asio::signal_set signals(main_service);
```

That says, we just need to create io_service_pool with value 1. (for node manager heartbeat handler) & just have another main_service on the main thread for both main service & signal service

rkooo567 replied:
EX) https://github.com/ray-project/ray/blob/d4324a4a8f886f60cd5f12356b97466e86a78b89/src/ray/raylet/main.cc#L246

rkooo567 replied:
In conclusion, we will only have 2 threads, main thread with main_service that is in charge of both signal & main logic + extra thread that is dedicating to node manager detctor

WangTaoTheTonic replied:
> I probably didn't understand here. If you use main_service on the main thread, there's no need to join them right? What I am saying here is
> 
> ```
> boost::asio::io_service main_service;
> ...
> boost::asio::signal_set signals(main_service);
> ```
> 
> That says, we just need to create io_service_pool with value 1. (for node manager heartbeat handler) & just have another main_service on the main thread for both main service & signal service

The thread will be join in stop method  of  io_service_pool.

rkooo567 replied:
In my example, it doesn't matter that `io_service_pool` joins inside signal handler. What I am saying is just to define 1 thread inside `io_service_pool`, which is just for NodeFailureDetector. And just keep the main_service for both signal & main thread logic. It should not cause deadlocks.

rkooo567 replied:
```cpp
// Use 1 thread for node failure detector event loop.
io_service_pool_ = std::make_shared<IOServicePool>(1);

// Share main_service for both signal & main event loop
boost::asio::io_service main_service;
...
boost::asio::signal_set signals(main_service);
```

WangTaoTheTonic replied:
> ```c++
> // Use 1 thread for node failure detector event loop.
> io_service_pool_ = std::make_shared<IOServicePool>(1);
> 
> // Share main_service for both signal & main event loop
> boost::asio::io_service main_service;
> ...
> boost::asio::signal_set signals(main_service);
> ```

This solution will work, but why don't we separate the main service and signal service but merge them together?

rkooo567 replied:
I think there's no need to add extra complexity (I think additional threads are always adding complexity) when it is not necessary.

WangTaoTheTonic replied:
@rkooo567 reasonable. Ockham's Razor rule. :)

WangTaoTheTonic commented:
> I think we've discussed last time, but it will definitely have synchronization issues. For example, `HandleHeartBeat` inside NodeFailureDetector (
> 
> https://github.com/ray-project/ray/blob/f2705e2c73605618dfa0b359447b4005bb63ac80/src/ray/gcs/gcs_server/gcs_node_manager.cc#L47
> 
> ) will be called by a main thread (in your implementation, it is not a main thread, but let's say it is). The methods used inside `Tick` will be used by node manager's io service thread. Both shares `heartbeat_buffer_` (
> https://github.com/ray-project/ray/blob/f2705e2c73605618dfa0b359447b4005bb63ac80/src/ray/gcs/gcs_server/gcs_node_manager.cc#L62
> 
> ). This means there could be some synchronization issues.
> Let me know if I am missing something, but it seems to me like all the states inside `NodeFailureDetector` should be protected by locks.

Thanks for pointing it out. But I don't think adding a mutex lock is a good idea as the lock will also block logic in node failure detector, which will delay the heartbeat check....
Optionally I put all the node failure detector actions into its io service by using post method. This will ensure the data structures will be accessed only in io service of failure detector.

rkooo567 commented:
@WangTaoTheTonic 

I am really concerned lock-free nature of `NodeFailureDetector` will cause synchronization issues so easily. Every public method of `NodeFailureDetector` should be posted to the event loop to avoid issues. It is really easy to do mistakes especially for new developers. 

Can we force developers to run all public methods of `NodeFailureDetector` inside the event loop? I am not sure what's the best solution here, but without this, I believe it is really prone to errors.

WangTaoTheTonic commented:
> @WangTaoTheTonic
> 
> I am really concerned lock-free nature of `NodeFailureDetector` will cause synchronization issues so easily. Every public method of `NodeFailureDetector` should be posted to the event loop to avoid issues. It is really easy to do mistakes especially for new developers.
> 
> Can we force developers to run all public methods of `NodeFailureDetector` inside the event loop? I am not sure what's the best solution here, but without this, I believe it is really prone to errors.

Actually i've tried the lock solution first. It ends up with acquiring lock in every(literally) method of node failure detector.....
I agree with you that even the main thread holds the lock, it will block detector thread not very long time.
Either is ok with me. I prefer current solution just because it is cleaner in codes, and there's also codes like this in core worker(https://github.com/ray-project/ray/blob/master/src/ray/core_worker/core_worker.cc#L377).

rkooo567 commented:
@WangTaoTheTonic Gotcha. Can you add some notes (with uppercase haha) that every public method should be posted to the event loop to avoid lock issues? I think that's the best thing we can do for now.

rkooo567 commented:
@WangTaoTheTonic Let me know after you addressed comments! I will take a look if tests failures are relevant.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Ray-PRB/28925/
Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Ray-PRB/28926/
Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Ray-PRB/28927/
Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Ray-PRB/28928/
Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Ray-PRB/28930/
Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Ray-PRB/28960/
Test FAILed.

WangTaoTheTonic commented:
> @WangTaoTheTonic Let me know after you addressed comments! I will take a look if tests failures are relevant.

@rkooo567 Comments addressed. Please take another look. The failed test should be irrelevant. Thanks~

rkooo567 commented:
@stephanie-wang Oh, that's a good catch. We definitely missed it. @WangTaoTheTonic Can you address that comment?

stephanie-wang commented on src/ray/common/constants.h:
```diff
@@ -42,3 +42,7 @@ constexpr char kWorkerRayletConfigPlaceholder[] = "RAY_WORKER_RAYLET_CONFIG_PLAC
 constexpr char kPublicDNSServerIp[] = "8.8.8.8";
 constexpr int kPublicDNSServerPort = 53;
+
+/// IO service used by gcs node failure detector, which avoid heartbeat dealy caused by
```
```suggestion
/// IO service used by gcs node failure detector, which avoid heartbeat delay caused by
```

stephanie-wang commented on src/ray/gcs/gcs_server/gcs_node_manager.cc:
```diff
@@ -203,7 +204,11 @@ void GcsNodeManager::HandleReportHeartbeat(const rpc::ReportHeartbeatRequest &re
   auto heartbeat_data = std::make_shared<rpc::HeartbeatTableData>();
   heartbeat_data->CopyFrom(request.heartbeat());
-  node_failure_detector_->HandleHeartbeat(node_id, *heartbeat_data);
+  // NOTE: TO AVOID HEARTBEAT BEING DEALYED BY MAIN THREAD, MAKE SURE HEARTBEAT IS ALWAYS
+  // HANDLED BY ITS OWN IO SERVICE
```
```suggestion
  // Note: To avoid heartbeats being delayed by main thread, make sure heartbeat is always
  // handled by its own IO service.
```

stephanie-wang commented on src/ray/gcs/gcs_server/gcs_node_manager.cc:
```diff
@@ -338,7 +343,11 @@ void GcsNodeManager::AddNode(std::shared_ptr<rpc::GcsNodeInfo> node) {
     RAY_CHECK(cluster_resources_.emplace(node_id, rpc::ResourceMap()).second);
     // Register this node to the `node_failure_detector_` which will start monitoring it.
-    node_failure_detector_->AddNode(node_id);
+    // NOTE: TO AVOID HEARTBEAT BEING DEALYED BY MAIN THREAD, MAKE SURE NODE ADDING IS
+    // ALWAYS HANDLED BY ITS OWN IO SERVICE
```
```suggestion
```suggestion
    // Note: To avoid heartbeats being delayed by main thread, make sure heartbeat is always
    // handled by its own IO service.
```
```

stephanie-wang commented on src/ray/gcs/gcs_server/gcs_node_manager.cc:
```diff
@@ -410,7 +419,11 @@ void GcsNodeManager::LoadInitialData(const EmptyCallback &done) {
 
-void GcsNodeManager::StartNodeFailureDetector() { node_failure_detector_->Start(); }
+void GcsNodeManager::StartNodeFailureDetector() {
+  // NOTE: TO AVOID HEARTBEAT BEING DEALYED BY MAIN THREAD, MAKE SURE NODE ADDING IS
+  // ALWAYS HANDLED BY ITS OWN IO SERVICE
```
```suggestion
```suggestion
  // Note: To avoid heartbeats being delayed by main thread, make sure node addition is always
  // handled by its own IO service.
```
```

stephanie-wang commented on src/ray/gcs/gcs_server/gcs_node_manager.h:
```diff
@@ -159,6 +159,9 @@ class GcsNodeManager : public rpc::NodeInfoHandler {
         std::shared_ptr<gcs::GcsPubSub> gcs_pub_sub,
         std::function<void(const ClientID &)> on_node_death_callback);
 
+    /// NOTE: TO AVOID HEARTBEAT BEING DEALYED BY MAIN THREAD, MAKE SURE EVERY PUBLIC
+    /// METHOD BELOW SHOULD BE POSTED TO ITS OWN IO SERVICE.
```
```suggestion
```suggestion
    // Note: To avoid heartbeats being delayed by main thread, all public methods below
    // should be posted to its own IO service.
```
```

WangTaoTheTonic commented:
> Please ping me again after addressing Stephanie's comment!

Addressed. CI pending to check.

P.S. The notes were transformed into lower case per Stephanie's comments...

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Ray-PRB/29042/
Test FAILed.
